wallet/psbt.go

```diff
@@ -0,0 +1,300 @@
+// must be the last signer of the transaction. That means, if there are any
+// unsigned non-witness inputs or inputs without UTXO information attached or
+// inputs without witness data that do not belong to the wallet, this method
+// will fail. If no error is returned, the PSBT is ready to be extracted and the
+// final TX within to be broadcast.
+//
+// NOTE: This method does NOT publish the transaction after it's been finalized
+// successfully.
+func (w *Wallet) FinalizePsbt(packet *psbt.Packet) error {
+	// Let's check that this is actually something we can and want to sign.
+	// We need at least one input and one output.
+	err := psbt.VerifyInputOutputLen(packet, true, true)
+	if err != nil {
+		return err
+	}
+
+	// Go through each input that doesn't have final witness data attached
+	// to it already and try to sign it. We do expect that we're the last
+	// ones to sign. If there is any input without witness data that we
+	// cannot sign because it's not our UTXO, this will be a hard failure.
+	tx := packet.UnsignedTx
+	sigHashes := txscript.NewTxSigHashes(tx)
+	for idx, txIn := range tx.TxIn {
+		in := packet.Inputs[idx]
+
+		// We can only sign witness inputs. Everything else must already
+		// have been signed or the finalize step later will fail.
+		if in.WitnessUtxo == nil {
+			continue
+		}
+
+		// Skip this input if it's got final witness data attached.
+		if len(in.FinalScriptWitness) > 0 {
+			continue
+		}
+
+		// We can only sign this input if it's ours, so we try to map it
+		// to a coin we own. If we can't, then we'll continue as it
+		// isn't our input.
+		info, _, err := w.FetchInputInfo(&txIn.PreviousOutPoint)
+		if err != nil {
+			continue
+		}
+
+		// Sanity check to avoid signing with something other than was
+		// provided in the PSBT.
+		if info.Value != in.WitnessUtxo.Value ||
+			!bytes.Equal(info.PkScript, in.WitnessUtxo.PkScript) {
+
+			return fmt.Errorf("found UTXO %#v but it doesn't "+
+				"match PSBT's input %v", info, in.WitnessUtxo)
+		}
+
+		// Now that we know the input is ours, we'll populate the
+		// signDesc with the per input unique information.
+		output := &wire.TxOut{
+			Value:    info.Value,
+			PkScript: info.PkScript,
+		}
+
+		// Finally, we'll sign the input as is, and populate the input
+		// with the witness and sigScript (if needed).
+		witness, sigScript, err := w.ComputeInputScript(
+			tx, output, idx, sigHashes, in.SighashType, nil,
+		)
+		if err != nil {
+			return fmt.Errorf("error computing input script for "+
+				"input %d: %v", idx, err)
+		}
+
+		// Serialize the witness format from the stack representation to
+		// the wire representation.
+		var witnessBytes bytes.Buffer
+		err = psbt.WriteTxWitness(&witnessBytes, witness)
+		if err != nil {
+			return fmt.Errorf("error serializing witness: %v", err)
+		}
+		packet.Inputs[idx].FinalScriptWitness = witnessBytes.Bytes()
+		packet.Inputs[idx].WitnessScript = sigScript
```

Comment:
This should be the witness script instead. It looks like things get by atm since this isn't a multi-sig output, and it only checks these fields for nil-ness when it goes to finalize it.

Reply:
Added a test case that uses a nested P2WKH input and that caught the problem. Should work correctly now.